### PR TITLE
Refactor AuthenticationController

### PR DESCRIPTION
**Pull request by yuya-yuliya**
- add AuthenticationManager
- refactor AuthenticationService

### CODE REVIEW

File: src/Feature/Account/code/Controllers/AuthenticationController.cs [hunk 1]
```diff
@@ -14,73 +14,54 @@
...
+            Assert.ArgumentNotNull(authenticationService, nameof(authenticationService));
+
             this.authenticationService = authenticationService;
-            this.cartService = cartService;
-            this.customerProvider = customerProvider;
-            this.commerceTrackingService = commerceTrackingService;
-            this.visitorContext = visitorContext;
         }
 
+        // TODO: Add error transfer to FE
```
Comment (larionovadaria) [inline src/Feature/Account/code/Controllers/AuthenticationController.cs]: - `Login(LoginRequest request \{email, password}\)`
- change only routing on FE (login/logout)
- add attribute validation in login request model

File: src/Feature/Account/code/Controllers/AuthenticationController.cs [hunk 2]
```diff
@@ -93,47 +74,25 @@ public ActionResult ValidateCredentials(UserLoginModel userLogin)
...
-            var anonymousContact = this.visitorContext.ContactId;
-            this.visitorContext.CurrentUser = user;
-
-            this.cartService.MergeCarts(anonymousContact);
-
-            this.commerceTrackingService.IdentifyAs("CommerceUser", user.UserName);
-        }
-
-        private bool LoginUser(UserLoginModel userLogin, out User user)
+        private bool ValidateUser(UserLoginModel userLogin)
```
Comment (larionovadaria) [inline src/Feature/Account/code/Controllers/AuthenticationController.cs]: Please move this one to service and mark as obsolete.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/Foundation/Account/code/Services/Authentication/AuthenticationService.cs [hunk 3]
```diff
@@ -14,24 +14,71 @@
...
+            var args = new LoginPipelineArgs()
+            {
+                Email = email,
+                Password = password
+            };
+
+            this.pipelineService.RunPipeline(Constants.Pipelines.Login, args);
+
+            var errorMessages = args.GetMessages(PipelineMessageFilter.Errors);
+            return !args.Aborted && errorMessages.Length > 0
```
Comment (stanislav-kozorez) [inline src/Foundation/Account/code/Services/Authentication/AuthenticationService.cs]: Do we actually have such situation when our pipeline is not aborted but we have error messages?
And do we actually need to have LoginResult now, when we can identify success or failure with `Result.Success` property?

Comment (yuya-yuliya) [reply src/Foundation/Account/code/Services/Authentication/AuthenticationService.cs]: We have discussion with Daria and rework this logic. Later we will update FE part and this information will be needed

File: src/Feature/Account/code/Constants.cs [hunk 4]
```diff
@@ -19,6 +19,7 @@ public static class Constants
         public static class Redirects
         {
             public const string SignIn = "/signIn";
```
Comment (larionovadaria) [inline src/Feature/Account/code/Constants.cs]: login?

File: src/Feature/Account/code/Models/Requests/LoginRequest.cs [hunk 5]
```diff
@@ -12,15 +12,21 @@
...
+    using System.Diagnostics.CodeAnalysis;
+
     using TypeLite;
 
+    [ExcludeFromCodeCoverage]
     [TsClass]
-    public class UserLoginModel
+    public class LoginRequest
     {
+        [Required]
```
Comment (larionovadaria) [inline src/Feature/Account/code/Models/Requests/LoginRequest.cs]: [EmailAddress]

File: src/Foundation/Account/code/Infrastructure/Pipelines/Login/LoginPipelineArgs.cs [hunk 6]
```diff
@@ -35,7 +35,7 @@ public class LoginPipelineArgs : PipelineArgs
 
         public string Password { get; set; }
 
-        public bool IsValid { get; set; }
+        public bool IsValidUser { get; set; }
```
Comment (larionovadaria) [inline src/Foundation/Account/code/Infrastructure/Pipelines/Login/LoginPipelineArgs.cs]: IsValidCredentials

File: src/Foundation/Account/code/Services/Authentication/IAuthenticationService.cs [hunk 7]
```diff
@@ -14,22 +14,38 @@
...
+        /// Validates user existence
         /// </summary>
-        /// <param name="userName">Name of the user.</param>
-        /// <param name="password">The password.</param>
-        /// <returns><c>true</c> if user was logged in, otherwise - <c>false</c>.</returns>
-        bool Login(string userName, string password);
+        /// <param name="email">User email</param>
+        /// <param name="password">User password</param>
+        /// <returns>True if user exists, otherwise - false</returns>
+        [Obsolete("Will removed")]
```
Comment (larionovadaria) [inline src/Foundation/Account/code/Services/Authentication/IAuthenticationService.cs]: Will be removed

File: src/Foundation/Account/code/Models/Authentication/LoginStatus.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+// 
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+namespace Wooli.Foundation.Account.Models.Authentication
+{
+    public enum LoginStatus
```
Comment (larionovadaria) [inline src/Foundation/Account/code/Models/Authentication/LoginStatus.cs]: to remove

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/Feature/Account/code/Controllers/AuthenticationController.cs [hunk 9]
```diff
@@ -14,126 +14,82 @@
...
-            this.visitorContext.CurrentUser = null;
-
-            this.commerceTrackingService.EndVisit(true);
-            this.Session.Abandon();
-            this.authenticationService.Logout();
-
-            return this.RedirectOnSignIn(null);
+            return this.Execute(
+                () => this.authenticationService.Logout(),
+                result => this.RedirectOnSignIn(null));
```
Comment (larionovadaria) [inline src/Feature/Account/code/Controllers/AuthenticationController.cs]: RedirectOnLogout/Login?

File: src/Feature/Account/code/Controllers/AuthenticationController.cs [hunk 10]
```diff
@@ -14,126 +14,82 @@
...
-            return this.RedirectOnSignIn(returnUrl);
+            return this.Execute(
+                () => this.authenticationService.Login(request.Email, request.Password),
+                result => 
+                {
+                    if (result.Success)
+                    {
+                        return result.Data != null && result.Data.IsValidUser
+                            ? this.RedirectOnSignIn(null)
+                            : this.JsonError("Invalid user", HttpStatusCode.Forbidden);
```
Comment (larionovadaria) [inline src/Feature/Account/code/Controllers/AuthenticationController.cs]: `this.JsonError(result.Error, HttpStatusCode.Forbidden)`

File: src/Feature/Account/code/Controllers/AuthenticationController.cs [hunk 11]
```diff
@@ -14,126 +14,83 @@
...
-
-            return this.RedirectOnSignIn(returnUrl);
+            return this.Execute(
+                () => this.authenticationService.Login(request.Email, request.Password),
+                result => 
+                {
+                    if (result.Success)
+                    {
+                        return result.Data != null && result.Data.IsValidCredentials
+                            ? this.RedirectOnAuthentication(null)
```
Comment (larionovadaria) [inline src/Feature/Account/code/Controllers/AuthenticationController.cs]: `this.RedirectOnAuthentication(returnUrl)`

**Review by at-nage — APPROVED** [review]

**Review by evgenii-nikulin — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**